### PR TITLE
Swap private implementation of tensor/pipeline parallel with using orchard

### PR DESCRIPTION
**Pull request by shaahji**
## Describe your changes
Swap private implementation of tensor/pipeline parallel with using orchard.

Orchard implements networks generically allowing Olive to provide only the distribution framework. Current example showcases the use of transformer network to implement both tensor parallelism and pipeline parallelism. The generated ranked models can be used with the driver script in the orchard repository.

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

**Comment by trajepl** [comment]
> Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

Should we update the examples page to show tensor/pipeline parallel example?

Also could you also help update the docs for the new pass?

**Comment by shaahji** [comment]
> > Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.
> 
> Should we update the examples page to show tensor/pipeline parallel example?
> 
> Also could you also help update the docs for the new pass?

The change is dependent on orchard repo which I am unsure whether or not @devang-ml wishes to publicize. Let me confirm it with him on Monday and will update the docs accordingly.

File: examples/orchard/user_script.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import torch
+
+
+def load_model(model_name):
+    from orchard.networks.transformer import Transformer
```
[the remark this answers is not in the document]
Comment (shaahji) [reply examples/orchard/user_script.py:10]: I have an open PR in that repo.

File: olive/passes/pytorch/tensor_parallel.py [hunk 2]
```diff
@@ -109,15 +85,9 @@ def _generate_one(params):
...
-        model_type = get_model_type_from_hf_config(hf_config)
-
-        if model_type == "llama":
-            from olive.passes.pytorch.tensor_parallel_llama2 import LlamaPyTorchTensorParallel
+        from olive.passes.pytorch.tensor_parallel_transformer import TransformerTensorParallel
 
-            impl = LlamaPyTorchTensorParallel(rank, world_size)
-        else:
-            raise ValueError("Unsupported model type '{model_type}' for tensor parallel pass")
+        impl = TransformerTensorParallel(rank, world_size)
```
Comment (guotuofeng) [inline olive/passes/pytorch/tensor_parallel.py:90]: do we need use factory to construct the implementation of TensorParallel?

File: olive/passes/pytorch/pipeline_parallel.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    @staticmethod
+    def _generate_one(params):
+        model_config, rank, world_size, output_filepath = params
+
+        logger.debug("Exporting pipeline parallel model for rank: %d, %s", rank, output_filepath)
+
+        from olive.passes.pytorch.pipeline_parallel_transformer import TransformerPipelineParallel
+
+        impl = TransformerPipelineParallel(rank, world_size)
```
Comment (guotuofeng) [inline olive/passes/pytorch/pipeline_parallel.py:86]: ditto

File: olive/passes/pytorch/pipeline_parallel.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    @abstractmethod
+    def restore_layers(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def split_layers(self, model: torch.nn.Module):
+        raise NotImplementedError
+
+
+class PyTorchPipelineParallel(Pass):
```
Comment (guotuofeng) [inline olive/passes/pytorch/pipeline_parallel.py:43]: it seems almost the logic is duplicated with that one in tensor_paralle.py except:
1. world_size input
2. generate PyTorchPipelineParallel or PyTorchTensorParallel
3. load rank weight or not.

Is it possible to make the redundancy code reusable?

File: olive/passes/pytorch/pipeline_parallel.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    @staticmethod
+    def _validate_nstages(v):
+        if int(v) < 2:
+            raise ValueError("nstages should be >= 2")
+
+        return v
+
+    @staticmethod
+    def _validators() -> Dict[str, Callable]:
```
Comment (guotuofeng) [inline olive/passes/pytorch/pipeline_parallel.py:71]: please rebase from main, the signature is changed to classmethod instead of staticmethod

**Comment by shaahji** [comment]
Orchard has been sidelined for the time being.

**Pull request closed without merging**